clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,914 @@
+                        throw new InterruptException(e);
+                    }
+                }
+
+                if (subscription == null) {
+                    log.debug("Client instance ID could not be retrieved with timeout {}", timeout);
+                    return Optional.empty();
+                }
+
+                Uuid clientInstanceId = subscription.clientInstanceId();
+                if (clientInstanceId == null) {
+                    log.info("Client instance ID was null in telemetry subscription while in state {}", state);
+                    return Optional.empty();
+                }
+
+                return Optional.of(clientInstanceId);
+            } finally {
+                lock.writeLock().unlock();
+            }
+        }
+
+        @Override
+        public void close() {
+            log.debug("close telemetry sender for client telemetry reporter instance");
+
+            maybeInitiateClose(Duration.ofMillis(MAX_TERMINAL_PUSH_WAIT_MS));
+
+            boolean shouldClose = false;
+            try {
+                lock.writeLock().lock();
+                if (this.state != ClientTelemetryState.TERMINATED) {
+                    if (maybeSetState(ClientTelemetryState.TERMINATED)) {
+                        shouldClose = true;
+                    }
+                } else {
+                    log.debug("Ignoring subsequent close");
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+
+            if (shouldClose) {
+                collectors.forEach(MetricsCollector::stop);
+            }
+        }
+
+        private Optional<Builder<?>> createSubscriptionRequest(ClientTelemetrySubscription localSubscription) {
+            // If we've previously retrieved a subscription, it will contain the client instance ID
+            // that the broker assigned. Otherwise, per KIP-714, we send a special "null" UUID to
+            // signal to the broker that we need to have a client instance ID assigned.
+            Uuid clientInstanceId = (localSubscription != null) ? localSubscription.clientInstanceId() : Uuid.ZERO_UUID;
+
+            try {
+                lock.writeLock().lock();
+                if (isTerminatingState()) {
+                    return Optional.empty();
+                }
+
+                if (!maybeSetState(ClientTelemetryState.SUBSCRIPTION_IN_PROGRESS)) {
+                    return Optional.empty();
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+
+            AbstractRequest.Builder<?> requestBuilder = new GetTelemetrySubscriptionsRequest.Builder(
+                new GetTelemetrySubscriptionsRequestData().setClientInstanceId(clientInstanceId), true);
+            return Optional.of(requestBuilder);
+        }
+
+        private Optional<Builder<?>> createPushRequest(ClientTelemetrySubscription localSubscription) {
+            if (localSubscription == null) {
```

Comment:
I would put this if block inside the lock because it does actually attempt a state change.

Reply:
The state change is protected by writelock hence I avoided same here. Please let me know if I am missing something here.

Reply:
This is true. It is safe. In terms of style, I prefer the entire method to be inside the write lock but happy for you to leave it as it is.